File: fabric-audit-agent/core/automation/dedupe.js
/**
 * Remove duplicate findings by key. Findings without a key are always kept
 * (so keyless test fakes are unaffected).
 * @param {object[]} findings
 * @returns {object[]}
 */
export function dedupe(findings) {
  const seen = new Set();
  const out = [];
  for (const f of findings) {
    if (f.key == null) { out.push(f); continue; }
    if (seen.has(f.key)) continue;
    seen.add(f.key);
    out.push(f);
  }
  return out;
}

File: fabric-audit-agent-app/fabric_audit_agent/automation/dedupe.py
"""Remove duplicate findings by key. Port of ``core/automation/dedupe.js``.

Findings without a key are always kept (keyless test fakes are unaffected).
"""


def dedupe(findings):
    seen = set()
    out = []
    for f in findings:
        if f.get("key") is None:
            out.append(f)
            continue
        if f["key"] in seen:
            continue
        seen.add(f["key"])
        out.append(f)
    return out
